content/blog/2025/07-01_httr2_client/index.qmd

```diff
@@ -0,0 +1,623 @@
+
+      - uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          extra-packages: |
+            any::rcmdcheck
+            any::knitr
+          needs: check
+
+      - name: "Setup env for full test suite"
+        run: |
+          echo "NETTSKJEMA_CLIENT_ID=${{ secrets.CLIENT_ID }}" >> .Renviron
+          echo "NETTSKJEMA_CLIENT_SECRET=${{ secrets.CLIENT_SECRET }}" >> .Renviron
+          echo "VCR_TURN_OFF=true" >> .Renviron
+
+      - uses: r-lib/actions/check-r-package@v2
+        with:
+          upload-snapshots: true
+          build_args: 'c("--no-manual","--compact-vignettes=gs+qpdf","--as-cran")'
+
+      - name: Build vignettes
+        run: |
+          install.packages(".", repos = NULL, type = "source")
+          nettskjemar:::knit_vignettes()
+        shell: Rscript {0}
+```
+
+## Making vignettes that call the API
+
+The last bit of the puzzle, and which you will see in my github action above, is that I call the function `nettskjemar:::knit_vignettes()`.
+This is an _internal_ function in my package that will knit my vignettes.
+
+"But why aren't your vignettes knit on check or build as normal?", you may ask.
+Because, the vignettes need the API credentials to knit.
+I've made them so that they do, and that I can make sure the vignette code actually work.
+
+How do I do that then?
+I take a note from the [rOpenSci blog](https://ropensci.org/blog/2019/12/08/precompute-vignettes/#the-solution-locally-knitting-rmarkdown) on locally knitting the vignettes.
+So rather than having a vignette called `nettskjemar.rmd` I have one called `nettskjemar.rmd.orig`. 
+By having this extra ending, R build will ignore the vignettes.
+We rather build the vignettes our selves, which I do with the handy internal function I made:
+
+```r
+knit_vignettes <- function() {
+  proc <- list.files(
+    "vignettes",
+    "orig$",
+    full.names = TRUE
+  )
+
+  lapply(proc, function(x) {
+    fig_path <- "static"
+    knitr::knit(
+      x,
+      gsub("\\.orig$", "", x)
+    )
+    imgs <- list.files(fig_path, full.names = TRUE)
+    sapply(imgs, function(x) {
+      file.copy(
+        x,
+        file.path("vignettes", fig_path, basename(x)),
+        overwrite = TRUE
+      )
+    })
+    invisible(unlink(fig_path, recursive = TRUE))
+  })
+
+  list(
+    "Knit vignettes",
+    sapply(proc, basename)
+  )
+} 
+```
+
+Ok, it's not that little, a lot is going on here.
+First, we locate all files ending with `orig` in the vignettes folder, and then we make sure the output name is without the extra `.orig` extension. 
+Then we also make sure to carry over any images the vignettes may create into the vignettes folder.
+Since, the knitting will happen in the project root, rather than in the vignettes folder, the output images will not be in the correct place.Making sure we carry over the output files is thus very important.
+
+I've found this code works well for this purpose, and I use it in a couple of my packages actually.
```

Comment:
if you use `usethis::use_release_issue()`, you can remind yourself to update the vignette https://usethis.r-lib.org/reference/use_release_issue.html#customization